virttest/qemu_vm.py

```diff
@@ -3526,13 +3527,14 @@ def create(self, name=None, params=None, root_dir=None,
             self.create_serial_console()
 
             for key, value in list(self.logs.items()):
-                outfile = "%s-%s.log" % (key, name)
+                outfile = os.path.join(utils_logfile.get_log_file_dir(),
+                                       "%s-%s.log" % (key, name))
                 self.logsessions[key] = aexpect.Tail(
                     "nc -U %s" % value,
                     auto_close=False,
-                    output_func=utils_misc.log_line,
+                    output_func=utils_logfile.log_line,
                     output_params=(outfile,))
-                self.logsessions[key].set_log_file(outfile)
```

Comment:
Hello @pevogam, I found a problem when destroying the VM, this behavior will clean the existing log file and renew one, causing old log contents missed. Here are serial log: `tail: io-github-autotest-qemu.boot.arm64-pci/serial-serial0-avocado-vt-vm1.log: file truncated`
```
2024-01-04 23:55:42: [  OK  ] Started System Logging Service.
2024-01-04 23:55:43: [   13.918355] block dm-0: the capability attribute has been deprecated.
tail: io-github-autotest-qemu.boot.arm64-pci/serial-serial0-avocado-vt-vm1.log: file truncated
2024-01-04 23:55:47: [   18.523651] sda2: Can't mount, would change RO state
2024-01-04 23:55:52: [   22.792339] audit: type=1305 audit(1704430552.379:164): op=set audit_pid=0 old=853 auid=4294967295 ses=4294967295 subj=system_u:system_r:auditd_t:s0 res=1
```

After the VM is created, serial session also opened to record all logs. However, destroy VM with graceful_shutdown will open a new serial session to send the shutdown command, so seems at that time, `file truncated` is triggered.

And I am curious that `Tail` can close the log file itself with `def _close_log_file`, why it needs `close_hooks` here? Does it have more benefits?

Reply:
Hi @PaulYuuu, `Tail` can no longer close log files and aexpect has a unified interface that only accepts an arbitrary function for handling any outputs. All the benefits from this were discussed in this pull request as well as https://github.com/avocado-framework/aexpect/commit/2d3154582f073a9b0cdcda44d1e3dbf67d1ff703#comments and the corresponding pull request on the aexpect side.

From your description it seems to me that we are not using all the right places to set up the logging since the above changes are meant to set logging only during vm creation and the vm graceful destruction is an unwanted side effect. In particular, we want to tie the logging of a session with the lifespan of a session and in turn to the lifespan of a vm. Can you point me to the exact lines that perform the unwanted truncation of the log file? IIUC the graceful shutdown of a VM is not supposed to create a new session, is that right? Do you have the exact line in question?

Reply:
Well, re-opening the file doesn't seem as an issue to me (eg. on double cleanup I can understand a log file reopenning), the real issue IMO sits in the `utils_logfile.py` where the file is opened as `w` where `a` would suite better (as we most certainly don't want to truncate the log files in any case.

Reply:
> Well, re-opening the file doesn't seem as an issue to me (eg. on double cleanup I can understand a log file reopenning), the real issue IMO sits in the `utils_logfile.py` where the file is opened as `w` where `a` would suite better (as we most certainly don't want to truncate the log files in any case.

I think this nailed it - all we need is the assumption of opening the log file multiple times. @PaulYuuu Do you agree with this? Are you willing to propose a PR or do you want me to do so?

Reply:
Thank you @ldoktor @pevogam, I agree with you, and I am sorry for my first comment, actually `close_own_log_file` is only called once. So the reason may be that when the fd is closed, some contents are still in the cache, so the process still logs them into the file, which reopens the file.

Reply:
@pevogam, I am glad to open a new one, let me try first.

Reply:
Hello @ldoktor @pevogam, I sent a new PR to fix this by change mode to 'a': https://github.com/avocado-framework/avocado-vt/pull/3822, thank you all. Cc @luckyh